org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend

```diff
@@ -21,32 +27,100 @@ import java.nio.charset.StandardCharsets
 @Singleton
 class UriExtensions {
 
+	static val ESCAPED_FILE_SCHEME = URLEncoder.encode("file:", StandardCharsets.UTF_8.name);
+
+	/**
+	 * Converts any non-{@code null} arguments, representing a URI as a string, into a EMF based URI where the following
+	 * rules apply:
+	 * <p>
+	 * <ul>
+	 * <li>If the argument has a {@code file} scheme:
+	 * <ul>
+	 * <li>and the {@code host} is not set (e.g.: {@code localhost} can be omitted) then it returns with a EMF URI with the
+	 * following form: {@code file:///path/to/resource}.</li>
+	 * <li>and the {@code host} is set (e.g.: {@code localhost} can be omitted) then drops the URI host and returns with a 
+	 * EMF URI with the following form: {@code file:///path/to/resource}.</li>
+	 * <li>if the URI starts with two forward-slashes ({@code file://path/to/resource}), the method will treat it with a URI
+	 * with {@code localhost} scheme and will omit it, producing the {@code file:///path/to/resource} result.</li>
+	 * </ul>
+	 * <li>If the has any other scheme then it will return with EMF URI {@link URI#createURI(String) created} from the
+	 * argument.</li>
+	 * <li>This method takes care of the encoding and the URI path escaping.</li>
+	 * </ul>
+	 * </p>
+	 */
 	def URI toUri(String pathWithScheme) {
-		// URI is used to get path of the uri without scheme
-		var path = URI.createURI(pathWithScheme).path
-		return URI.createURI(path.toPath)
+		// This will take care of the encoding.
+		val uri = URI.createURI(pathWithScheme.checkNotNull('pathWithScheme'));
+		if (uri.file) {
+			val segments = uri.toSegments;
```

Comment:
I am a bit afraid of manual parsing/serialization, we could lose some info from URIs, like query params

Reply:
Although I totally agree with you, I am positive the current state [does not handle this either](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend#L26).

The ball is in your court...

Reply:
ok, the state before as well